### ISSUE TITLE
Please allow customization of JSON key ordering when serializing dictionaries.

### ISSUE THREAD

**Issue by rummelsworth** [JamesNK/Newtonsoft.Json#2270]
### Source type

```csharp
Dictionary<string, int> // for example, with ("a", 1), ("c", 2), and ("b", 3), added in that order.
```

### Desired JSON

```javascript
{"a":1,"b":3,"c":2}
```

### Actual JSON

```javascript
{"a":1,"c":2,"b":3}
```

### Steps to reproduce

```csharp
void Main() // in linqpad
{
    var d = new Dictionary<string, int>
    {
        ["a"] = 1,
        ["c"] = 2,
        ["b"] = 3,
    };
    var s = new JsonSerializerSettings
    {
        ContractResolver = new OrderedPropertyNamesContractResolver(StringComparer.OrdinalIgnoreCase),
    };
    JsonConvert.SerializeObject(d).Dump(); // yields {"a":1,"c":2,"b":3}, as expected.
    JsonConvert.SerializeObject(d, s).Dump(); // yields the same, but was hoping for {"a":1,"b":3,"c":2}
}

// My very naive attempt at doing this without constructing a whole new SortedDictionary or SortedList ---
class OrderedPropertyNamesContractResolver : DefaultContractResolver
{
    readonly IComparer<string> Comparer;

    public OrderedPropertyNamesContractResolver(IComparer<string> comparer) => Comparer = comparer ?? throw new ArgumentNullException(nameof(comparer));

    protected override IList<JsonProperty> CreateProperties(Type type, MemberSerialization memberSerialization)
    {
        var properties = base.CreateProperties(type, memberSerialization).OrderBy(property => property.PropertyName, Comparer).ToList();
        foreach (var (property, index) in properties.Select(ValueTuple.Create<JsonProperty, int>))
        {
            property.Order = index;
        }
        return properties;
    }
}
```

### Current workaround

Construct a `SortedDictionary` or `SortedList` with the same entries as the source dictionary, and then serialize that. Possibly insufficient when the dictionary is actually inside an aggregate that you're serializing.

**Comment by bartelink** [comment]
Dictionaries are by definition arbitrarily ordered - a special case like this makes no sense to bake into a general library

**Comment by rummelsworth** [comment]
> [Dictionary keys] are by definition arbitrarily ordered

Agreed. So are POCO properties. Yet it makes lots of sense that custom ordering of the latter is allowed via `JsonPropertyAttribute.Order` and (e.g. within a custom contract resolver) `JsonProperty.Order`.

> a special case like this makes no sense to bake into a general library

Elaboration on this point would be appreciated.

**Comment by bartelink** [comment]
It's not a thing in the json spec.

Json.Net is a library thats ~10 years old. You have a workaround for your special case. It's pretty difficult to justify adding this case to library thats this mature and has survived this long without the special casing at this point.

But: once again: if a low level json writer was arbitrarily reordering items, that would be a problem. This is a Dictionary - I'd be very shocked and surprised that there's a way to order them.

TL;DR This is a ["starts with -100 points" issue](https://blogs.msdn.microsoft.com/ericgu/2004/01/12/minus-100-points/)

**Comment by rummelsworth** [comment]
> TL;DR This is a "starts with -100 points" issue

The implication being that some issues do not start with -100 points. Contrast this with what you linked:

> So, we decided on the additive approach instead, and worked hard to keep the complexity down. One way to do that is through the concept of “minus 100 points”. **Every feature starts out in the hole by 100 points**, which means that it has to have a significant net positive effect on the overall package.

So again, I think we agree. This feature, like all features, starts with -100 points.

> It's not a thing in the json spec.

Correct, I think. Not sure how this is relevant, though, since JSON.net allows custom ordering of JSON object keys when the keys come from .NET object properties.

> Json.Net is a library thats ~10 years old. You have a workaround for your special case. It's pretty difficult to justify adding this case to library thats this mature and has survived this long without the special casing at this point.

Again, mostly correct. It is about 14 years old. I have a partial workaround. The feature is indeed probably difficult to justify.

> if a low level json writer was arbitrarily reordering items, that would be a problem.

So... it's a problem that the current API allows customization of JSON object key ordering for .NET object properties. I disagree here, but it's a perfectly valid opinion for you to hold.

> This is a Dictionary - I'd be very shocked and surprised that there's a way to order them.

I'm sincerely trying to stand in your shoes on this, but I don't understand your "shock & surprise" here:

- Things with comparers can be ordered.
- Strings have comparers.
- JSON object keys are strings.
- JSON.net provides custom ordering when those keys are sourced from .NET object properties.
- JSON.net does not provide custom ordering when those keys are sourced from .NET dictionary types.

To be clear, I make no claim that this feature is easy to (a) implement or (b) integrate without breaking changes. I hope that it is both of those things, but if wishes were horses, beggars would ride.

---

I'm removing the "please" from the issue title, but not because I'm not a beggar. I am a beggar, as is almost any FOSS user making any request of the FOSS project they're using. I think the "please" implies I'm unwilling to take this on myself. I am willing. If I manage to schedule some time for it, I'll mention it here, just as a heads-up for other potential implementors/collaborators.

**Comment by bartelink** [comment]
I was being a devil's advocate - I won't be implementing the feature. I like software I use to be and stay simple. I also like maintainers not to have to ask devils advocate questions that others could ask. I consider that a public service.

I don't think your feature is completely crazy but
a) _all_ features start with -100 
b) Dictionaries, no matter how comparable their keys or vaales *are not ordered* 
c) this is not a systemic issue affecting lots of cases - its a non-idiomatic JSON thing that you have a workaround for

I'm not suggesting you're crazy or this is unimplementable, but having a general library support a very specific rendering like this smells funny to me.

That's all, no harm intended.

**Comment by rummelsworth** [comment]
- (a) Agreed, as mentioned above.
- (b) Agreed, but irrelevant as discussed above.
- (c) Agreed, but the workaround is partial, as mentioned above.

No offense taken. I think it's great you engaged with this. At the very least, our discussion has been a "smoke test" of the feature's reasonableness. Thanks.

**Comment by AlexSnowLeo** [comment]
Hi @rummelsworth and @bartelink.
You will be surprised, but I recently came across a similar problem. It is not concerned with sorting, but it stems from the fact that the basic method CreateProperties does not return properties for the type of dictionary. 
For example: 
- If I use Dictionary with determined value type 'int' CreateProperties call, but the basic method does not return anything.
- If I use Dictionary with Object value type CreateProperties method **is not called**.
This is expected behavior?

**Comment by robheffo79** [comment]
I personally would like to see this added. I am using Json serialisation to produce a "normalised" versions of arbitrary objects. Being able to ensure Dictionarys have consistantly ordered keys helps greatly because a dictionary doesn't care about the order, just that a key is present or not, whereas the normalisation process does care about both the order and whether or not a key is present..

**Comment by CZEMacLeod** [comment]
I've put together a little implementation for you... This is not very optimised and there may be a few differences from how you want it to work (although you should be able to tweak it from there).

While this should handle any generic type of `IDictionary<string,?>`, I haven't done anything around handling mixed dictionary with additional properties type objects, nor supporting things like `TypeNameHandling` and so on. I'm sure it could be fleshed out if required.

The code uses reflection to bind the methods, and that code should probably be made more thread-safe (although I think it should work pretty well as-is). There might be better ways to implement that using code expressions, but for a proof of concept I think it is good enough.

This is designed for net8.0, but if you strip the nullability stuff out, it should be fine on net4x too. You might need to tweak the code format as it uses collection initialisers in some places...

HTH

```cs
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using System.Collections;
using System.Diagnostics.CodeAnalysis;
using System.Reflection;

public static class Program
{
    public static void Main()
    {
        var d = new Dictionary<string, int>
        {
            ["a"] = 1,
            ["c"] = 2,
            ["b"] = 3,
        };
        var d2 = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["x"] = "World",
            ["Y"] = "Hello",
        };
        var a = new
        {
            values = d,
            Stuff = "this",
            greeting = d2
        };
        var s = new JsonSerializerSettings
        {
            ContractResolver = new SortedPropertyNamesContractResolver(StringComparer.Ordinal),
            Converters = [new SortedDictionaryConverter(StringComparer.Ordinal)]    // N.B. The sort order here is only used if the dictionary is an IDictionary and not a Dictionary with a sort order set
        };
        Console.WriteLine(JsonConvert.SerializeObject(a));    // yields {"values":{"a":1,"c":2,"b":3},"Stuff":"this","greeting":{"x":"World","Y":"Hello"}}
        Console.WriteLine(JsonConvert.SerializeObject(a, s)); // yields {"Stuff":"this","greeting":{"x":"World","Y":"Hello"},"values":{"a":1,"b":3,"c":2}}

        Console.WriteLine(JsonConvert.SerializeObject(d2, s)); // yields {"x":"World","Y":"Hello"}
        Console.WriteLine(JsonConvert.SerializeObject(new CustomDictionary<string>(d2), s)); // yields {"Y":"Hello","x":"World"}
        // CustomDictionary<string> is an IDictionary<string,string> but has no access to the key sort order, so we use the default passed to SortedDictionaryConverter
    
        var d3 = JsonConvert.DeserializeObject<Dictionary<string, string>>(JsonConvert.SerializeObject(d2,s), s); // Check we can deserialize
        foreach (var k in d3.Keys)
        {
            Console.WriteLine($"{k}={d3[k]}");
        }
    }
}

public class SortedDictionaryConverter : JsonConverter
{
    private readonly IComparer<string> comparer;

    public SortedDictionaryConverter() : this(StringComparer.Ordinal) { }
    public SortedDictionaryConverter(IComparer<string> comparer) => this.comparer = comparer ?? throw new ArgumentNullException(nameof(comparer));

    private static readonly Type idic = typeof(IDictionary<,>);

    public override bool CanConvert(Type objectType)
    {
        return objectType.GetInterfaces().Any(i => i.IsGenericType && i.GetGenericTypeDefinition() == idic &&
            i.GetGenericArguments().First() == typeof(string));
    }

    public override object ReadJson(JsonReader reader, Type objectType, object? existingValue, JsonSerializer serializer) => throw new NotImplementedException();

    public override bool CanWrite => true;
    public override bool CanRead => false;

    private static readonly Dictionary<Type, Action<JsonWriter, object, JsonSerializer, IComparer<string>>> writeJsonMethods = [];
    private static readonly MethodInfo wji = typeof(SortedDictionaryConverter).GetMethod(nameof(WriteJsonInternal), BindingFlags.Static | BindingFlags.NonPublic)!;

    public override void WriteJson(JsonWriter writer, object? value, JsonSerializer serializer)
    {
        if (value is null) { writer.WriteNull(); return; }
        if (value is IDictionary d && d.Count == 0) { writer.WriteStartObject(); writer.WriteEndObject(); return; }

        var iface = value.GetType().GetInterfaces().First(i => i.GetGenericTypeDefinition() == idic &&
            i.GetGenericArguments().First() == typeof(string));

        var dtype = iface.GetGenericArguments()[1]; // iface is IDictionary<string,{dtype}>

        if (!writeJsonMethods.TryGetValue(dtype, out var method))
        {
            var mi = wji.MakeGenericMethod(dtype) ?? throw new NotImplementedException();
            // 
            method = wji.MakeGenericMethod(dtype).CreateDelegate<Action<JsonWriter, object, JsonSerializer, IComparer<string>>>(null);
            writeJsonMethods[dtype] = method;
        }
        method(writer, value, serializer, comparer);
    }

    private static void WriteJsonInternal<T>(JsonWriter writer, object value, JsonSerializer serializer, IComparer<string> defaultComparer)
    {
        if (value is not IDictionary<string,T> idict) throw new ArgumentOutOfRangeException(nameof(value));

        var dict = idict as Dictionary<string, T>;
        var comparer = dict?.Comparer as IComparer<string> ?? defaultComparer; // Try and use the dictionary's comparer but fallback if unavailable

        writer.WriteStartObject();

        foreach (var key in idict.Keys.OrderBy(k => k, comparer))
        {
            writer.WritePropertyName(key);
            serializer.Serialize(writer, idict[key]);
        }

        writer.WriteEndObject();
    }
}

public class SortedPropertyNamesContractResolver : DefaultContractResolver
{
    private readonly IComparer<string> comparer;

    public SortedPropertyNamesContractResolver() : this(StringComparer.Ordinal) { }
    public SortedPropertyNamesContractResolver(IComparer<string> comparer) => this.comparer = comparer ?? throw new ArgumentNullException(nameof(comparer));

    protected override IList<JsonProperty> CreateProperties(Type type, MemberSerialization memberSerialization) => 
        base.CreateProperties(type, memberSerialization).OrderBy(property => property.PropertyName!, comparer).ToList();
    // This does not override explicitly ordered properties
}

public class CustomDictionary<T> : IDictionary<string,T>
{
    private readonly IDictionary<string, T> source;

    public CustomDictionary(IDictionary<string, T> source) => this.source = source;

    public T this[string key] { get => source[key]; set => source[key] = value; }

    public ICollection<string> Keys => source.Keys;

    public ICollection<T> Values => source.Values;

    public int Count => source.Count;

    public bool IsReadOnly => source.IsReadOnly;

    public void Add(string key, T value) => source.Add(key, value);

    public void Add(KeyValuePair<string, T> item) => source.Add(item);

    public void Clear() => source.Clear();

    public bool Contains(KeyValuePair<string, T> item) => source.Contains(item);

    public bool ContainsKey(string key) => source.ContainsKey(key);

    public void CopyTo(KeyValuePair<string, T>[] array, int arrayIndex) => source.CopyTo(array, arrayIndex);

    public IEnumerator<KeyValuePair<string, T>> GetEnumerator() => source.GetEnumerator();

    public bool Remove(string key) => source.Remove(key);

    public bool Remove(KeyValuePair<string, T> item) => source.Remove(item);

    public bool TryGetValue(string key, [MaybeNullWhen(false)] out T value) => source.TryGetValue(key, out value);

    IEnumerator IEnumerable.GetEnumerator() => ((IEnumerable)source).GetEnumerator();
}
```

**Comment by tsornin** [comment]
@CZEMacLeod I wanted to do this and your code helped a lot, thanks!

Two things that didn't work for me out of the box:

1. The interface get in `WriteJson` also needs the `i.IsGenericType &&` check
```csharp
var iface = value.GetType().GetInterfaces().First(i => i.IsGenericType && i.GetGenericTypeDefinition() == idic &&
    i.GetGenericArguments().First() == typeof(string));
```

2. My .NET version didn't have the generic `CreateDelegate`. I did this instead:
```csharp
if (!_writeJsonMethods.TryGetValue(tValueType, out var method))
{
    var genericWriteJsonInternalMethod = _writeJsonInternalMethod.MakeGenericMethod(tValueType);
    method = (Action<JsonWriter, object, JsonSerializer, IComparer<string>>)Delegate.CreateDelegate(typeof(Action<JsonWriter, object, JsonSerializer, IComparer<string>>), genericWriteJsonInternalMethod);
    _writeJsonMethods[tValueType] = method;
}
```
